name: Test

concurrency:
  group: ${{ github.head_ref || github.run_id }}
  cancel-in-progress: true

on:
  push:
  pull_request:
  workflow_dispatch:

env:
  FOUNDRY_PROFILE: ci

jobs:
  test:
    name: Foundry project
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
        with:
          submodules: recursive

      - uses: bgd-labs/action-rpc-env@main
        with:
          ALCHEMY_API_KEY: ${{ secrets.ALCHEMY_API_KEY }}

      - name: Run Foundry setup
        uses: bgd-labs/github-workflows/.github/actions/foundry-setup@main
        with:
          ZKSYNC: "true"

      - name: Show Forge version
        run: |
          forge --version

      - name: Run Forge fmt
        run: |
          forge fmt --check
        id: fmt

      - name: Run Forge tests
        id: test
        uses: bgd-labs/github-workflows/.github/actions/foundry-test@main

      - name: Run ZK tests
        id: zktest
        uses: bgd-labs/github-workflows/.github/actions/foundry-test@main
        with:
          ZKSYNC: true
          ROOT_DIR: "zksync"

      - name: Run Forge tests
        uses: bgd-labs/github-workflows/.github/actions/comment-artifact@main

      # we let failing tests pass so we can log them in the comment, still we want the ci to fail
      - name: Post test
        if: ${{ steps.test.outputs.testStatus != 0 || steps.zktest.outputs.testStatus != 0 }}
        run: |
          echo "tests failed"
          exit 1
